example-app/src/main/java/com/adyen/checkout/example/ui/instant/InstantFragment.kt

```diff
@@ -0,0 +1,139 @@
+import com.adyen.checkout.redirect.RedirectComponent
+import com.google.android.material.bottomsheet.BottomSheetDialogFragment
+import dagger.hilt.android.AndroidEntryPoint
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+
+@AndroidEntryPoint
+class InstantFragment : BottomSheetDialogFragment() {
+
+    @Inject
+    internal lateinit var checkoutConfigurationProvider: CheckoutConfigurationProvider
+
+    private var _binding: FragmentInstantBinding? = null
+    private val binding: FragmentInstantBinding
+        get() = requireNotNull(_binding)
+
+    private var instantPaymentComponent: InstantPaymentComponent? = null
+
+    private val instantViewModel: InstantViewModel by viewModels()
+
+    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
+        // Insert return url in extras, so we can access it in the ViewModel through SavedStateHandle
+        val returnUrl = RedirectComponent.getReturnUrl(requireActivity().applicationContext)
+        arguments = bundleOf(RETURN_URL_EXTRA to returnUrl)
+        _binding = FragmentInstantBinding.inflate(inflater, container, false)
+        return binding.root
+    }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        viewLifecycleOwner.lifecycleScope.launch {
+            repeatOnLifecycle(Lifecycle.State.STARTED) {
+                launch { instantViewModel.paymentMethodFlow.collect(::setupInstantComponent) }
+                launch { instantViewModel.events.collect(::onEvent) }
+                launch { instantViewModel.instantViewState.collect(::onViewState) }
+            }
+        }
+    }
+
+    private fun setupInstantComponent(paymentMethod: PaymentMethod) {
+        val instantPaymentComponent = InstantPaymentComponent.PROVIDER.get(
+            this,
+            paymentMethod,
+            checkoutConfigurationProvider.getInstantConfiguration(),
+            requireActivity().application,
+        )
+
+        this.instantPaymentComponent = instantPaymentComponent
+        instantPaymentComponent.observe(viewLifecycleOwner, instantViewModel::onPaymentComponentEvent)
+        binding.componentView.attach(instantPaymentComponent, viewLifecycleOwner)
+    }
+
+    private fun onEvent(event: InstantEvent) {
+        when (event) {
+            is InstantEvent.AdditionalAction -> {
+                onAction(event.action)
+            }
+            InstantEvent.Invalid -> {
+                // no ops
+            }
+            is InstantEvent.PaymentResult -> {
+                onPaymentResult(event.result)
+            }
+        }
+    }
+
+    private fun onPaymentResult(result: String) {
+        Toast.makeText(requireContext(), result, Toast.LENGTH_SHORT).show()
+        dismiss()
+    }
+
+    private fun onViewState(viewState: InstantViewState) {
+        when (viewState) {
+            InstantViewState.Error -> {
```

Comment:
can be something like that
```
    private fun handleViewState(
        isShowComponent: Boolean = false,
        isLoading: Boolean = false,
        isError: Boolean = false
    ) {
        binding.errorView.isVisible = isError
        binding.progressIndicator.isVisible = isLoading
        binding.componentContainer.isVisible = isShowComponent
    }
```
and just call the method with the right parameter inside the when clause 
`handleViewState(isShowComponent = true)`

Reply:
I think this reduces readability and makes it hard to understand what views we're hiding or displaying. Better to explicitly set the visibility of all the views we need to.